### ISSUE TITLE
Allow specifying the precision?

### ISSUE THREAD

**Issue by scotttrinh** [sarahdayan/dinero.js#13]
Any plan or desire to allow specifying the precision with `2` being the default? Seems useful to account for regulatory requirements or even just personal preference. I'm imagining something like this:

```js
Dinero({ amount: 1000, currency: 'USD' })
  .toFormat('$0,0.00'); // '$10.00'
Dinero({ amount: 10000000, precision: 6, currency: 'USD' })
  .toFormat('$0,0.00'); // '$10.00'
```

I'd be happy to PR this!

**Comment by sarahdayan** [comment]
There's a [project](https://github.com/sarahdayan/dinero.js/projects/3) for this. You can click "Menu" to see the description. I'm assigning it to this issue.

I feel like this deserves a discussion before starting to code 😊

**Comment by scotttrinh** [comment]
Awesome. Do you have an idea about how to have this discussion? Just talk about it on this issue or set up a call?

I can throw together a throw-away PR that has the most naivë implementation I can think of as a starting point for a discussion if that helps materialize some ideas and concerns.

**Comment by sarahdayan** [comment]
There are two approaches for adding the concept of precision: **added to**, or **baked in** the [exponent](https://en.wikipedia.org/wiki/ISO_4217#Treatment_of_minor_currency_units_(the_%22exponent%22)).

_For the record, the exponent is the factor by which you multiply 10 to get the number of sub-units in a unit. For dollars, the exponent is `2` as `10^2 = 100`._

Dinero.js doesn't support exponents properly yet, [but will in v2](https://github.com/sarahdayan/dinero.js/projects/5). Now even though it doesn't currently have the concept of an exponent, it implicitly does by dividing by `100`. This implicit behavior was reflected in [v1.1.2](https://github.com/sarahdayan/dinero.js/releases/tag/v1.1.2) to prepare for future support.

The exponent defacto indicates the default number of possible decimal places for a _billable_ amount. Yet you may want to represent prices with more precision:

- unit price for items sold in higher quantity, like stuff in bulk or gas
- avoid rounding during calculations to preserve precision

So this concept, loosely named precision here, aims at "adding" support for these fractional units so they can be represented and calculated.

Back to the two approaches, here how you can think about it:

### Solution 1: Precision is added to the exponent

In this case, precision starts at 0. Anything more adds up to the existing exponent of the currency.
Whatever the exponent is, precision adds up decimal places.

```js
Dinero({ amount: 104545, currency 'USD' })
  .toUnit() // 1045.45

Dinero({ amount: 104545, currency 'USD', precision: 2 })
  .toUnit() // 10.4545, or 104545 / 10^2+2
```

### Solution 2: Precision is "baked in" the exponent

In this case, precision starts at the existing exponent of the currency. Giving a precision that's equal to the exponent is like not giving a precision at all. Giving less throws an error, and giving more adds starting from 0.

```js
Dinero({ amount: 104545, currency 'USD' })
  .toUnit() // 1045.45

Dinero({ amount: 104545, currency 'USD', precision: 2 })
  .toUnit() // 1045.45, or 104545 / 10^2+2-2
```

Both solutions have advantages and flaws. I'm leaning towards **solution 2**, here's why.

I think it's easier to understand for the end user, and more straight-forward. When you see `precision: 2`, you expect two decimal places. If you have to always handle the same amount of decimal places even regardless of the currency, you can do it easily.

The only "flaw" I can think of is that adding a precision that's equal to the existing component does nothing, but in the end, if you have to use `precision` that's because you need more than what `exponent` already gives you.

I think solution 1 is confusing.

```js
Dinero({
  amount: 104541,
  currency 'USD',
  precision: 4
})
```

If you see the above piece of code, you have no idea what the end price is, and it gets worse as you need more decimal places. It's definitely not straight-forward that the above object represents $0.104541. If you're new on a project, you will need some help figuring out the underlying mechanism. If you need to build a new object that represents an amount with support for more decimal places than it does by default, you'll need to do some mental calculations that can become annoying and cumbersome in the long run.

I'm 100% open to debate, don't hesitate to jump in if you have converging or diverging opinions 👍

**Comment by scotttrinh** [comment]
Thanks @sarahdayan for such a thorough overview of the two solutions. I believe I have a good case for preferring Solution 1, so I'll add it here to further the debate.

There are only a few possibilities I can think of for needing addtional precision beyond the sub-unit of the currency.

1. Expressing a monetary value for something that is usually in bulk, like flour, gasoline, or a financial instrument like a future. You've already pointed out this use case.
2. Regulatory requirements. You might be required by some regulatory body to calculate to a certain fraction of the sub-unit.
3. You have a non-regulatory business requirement that requires or suggests extra precision when doing calculations.
4. You anticipate converting between currencies and need extra precision.

I'm sure there are a few others, but my guess (based on a survey of other implementations) is that this feature is not widely needed or used.

Given those use cases, my feeling is that the precision that you really care about is *additional precision beyond the sub-unit*. As you pointed out already, Solution 2 has a case where you can specify a value and it's the same as not specifying a value. That's not a big problem. I see throwing a runtime exception for specifying a smaller precision than the exponent as more problematic, personally.

If we can agree that it is not useful to set the precision to the value of the exponent, than the only meaningful values are positive integers greater than the exponent. Furthermore, all four cases suggest to me that the added precision is a fixed value and does not depend on the particular currency. Certainly that is just my impression of the problem-space and not a requirement. If we believe in my assumption about fixed-precision, I believe it suggests Solution 1.

On to the objection, which is totally valid, that the user will not be able to convert from the `Dinero.amount` back to the actual monetary unit in their heads: I don't think people will be dealing with the raw `.amount` property much directly.

It is very likely that in an actual system you will not be entering in `.amount` values directly from a provided variable or a `Number` literal. That assumes that the storage of these values is anticipating the exponent already, so likely you'll have some sort of function that converts from whatever your initial representation is to the properly padded `.amount` value. Something like:

```js
function fromDatabaseNumberToAmount(dbNum, exponent) {
  return dbNum * Math.pow(10, exponent);
}
```

If you're use case requires extra precision, you can easily just add your precision in this function:

```js
function fromDatabaseNumberToAmount(dbNum, exponent, precision) {
  return dbNum * Math.pow(10, exponent + precision);
}
```

Where does the exponent come from? That either needs to be communicated by the source of the amount in the first place (database, server, in-memory value) or we need to provide a way to query for the exponent.

Another idea is to provide helper functions to convert from format strings or the unit numbers back into `Dinero`:

```js
const item = db.getById('itemId'); // { price: '1,000.45' }
const dString = Dinero.fromString(item.price, '0,0.0', { currency: 'USD', precision: 4 });
dString.getAmount(); // 1000450000
dString.getUnit(); // 1000.45
```

```js
const item = db.getById('itemId'); // { price: 1000.45 } 
const dUnit = Dinero.fromUnit(item.price, { currency: 'USD', precision: 4 });
dUnit.getAmount(); // 1000450000
dUnit.getUnit(); // 1000.45
```

This takes the ambiguity out of it, and allows us to keep the exponent internal to the library. In Solution 2, you have to know the exponent in order to set the additional precision and not get a runtime exception. I believe the only time you need to reason about the `.amount` literal is during debugging, and it's easy enough to call `#toUnit` on the instance to take the guess work out of it. FWIW, you have to know the exponent when reasoning about the `.amount`, too.

**Comment by scotttrinh** [comment]
I definitely can sympathize with the idea that the name `precision` makes it seem like it should *include* the subunit (like USD is precision 2, JPY is precision 0), so I think we need to maybe consider another name if we go with Solution 1.

**Comment by sarahdayan** [comment]
Thanks a lot for your input @scotttrinh. Here are my thoughts about it.

I agree that `precision` isn't an ideal name. It's ambiguous and no matter what solution is picked, it needs to change accordingly. In solution 1, it's more a concept of `additionalSubUnitFractions`, while in solution 2, this would be `subUnitFractions`. I'm not settled on these terms but let's use them for the sake of clarity in the debate.

> If we can agree that it is not useful to set the precision to the value of the exponent, then the only meaningful values are positive integers greater than the exponent.

I agree with this, but I feel like how it translates in terms of code isn't satisfying because it creates serious cognitive overhead.

> All four cases suggest to me that the added precision is a fixed value and does not depend on the particular currency.

The advantage of additional sub-unit fractions is that you can always add the same amount of _additional_ decimal places, regardless of the currency. It's always "plus x". The inconvenience is that, if you want a specific number of overall decimal places, you need to do some mental math.

### Solution 1

I want 4 decimal places, I need to know what is the exponent and do `exponent - 4` to know `additionalSubUnitFractions`.

```js
// For an amount of 1
Dinero({ amount: 10000, currency: 'USD', additionalSubUnitFractions: 2 })
Dinero({ amount: 100000, currency: 'IQD', additionalSubUnitFractions: 1 })
```
### Solution 2

I want 4 decimal places, I can simply indicate it.

```js
// For an amount of 1 dollar/dinar
Dinero({ amount: 10000, currency: 'USD', subUnitFractions: 4 })
Dinero({ amount: 100000, currency: 'IQD', subUnitFractions: 4 })
```

## The issue of fluctuating exponents

Currently, `exponent` equals to `2`, all the time. It means that if you want to use the concept of precision to circumvent it, this is how you'd do it right now:

### Solution 1

```js
// For an amount of 1 dinar
Dinero({ amount: 1000, currency: 'IQD', additionalSubUnitFractions: 1 })
```

### Solution 2

```js
// For an amount of 1 dinar
Dinero({ amount: 1000, currency: 'IQD', subUnitFractions: 3 })
```

As exponent support is added, Dinero objects created with solution 2 will keep on working. All you'll have to do is remove the unnecessary `subUnitFractions` parameter, but it won't matter if you don't. With solution 1, you'll need to do a major migration. In the example above, your Dinero object representing 1 dinar will suddenly represent 0.1 dinar.

I know that all upgrades to a new major version involves migrating your app, but this one would definitely be annoying.

In addition to that, we can't guarantee that `exponent` won't change. If Iraqi dinar, that currently still officially use fils, but that inflation has made obsolete, finally drops sub-unit support (like Japanese Yens), `exponent` will shift from `3` to `0`. With solution 1, this will mess up all existing implementations.

The thing is, we shouldn't deprecate the API every time the currency dataset is changed, just because the implementation might suffer. **With solution 1, this will be inevitable**. With solution 2, instead of throwing when the precision is lower than the exponent (for example, if exponent goes from `0` to `3` and you had specified a precision of 2), you can just force the precision to be at least as much as the exponent, regardless of what the user specified. An additional warning could be thrown in dev mode but nothing would break.

I believe solution 2 isn't only clearer, it's also more robust and future-proof.

**Comment by joelnet** [comment]
Also consider large precision numbers for virtual currencies. Bitcoin has a precision of 8 decimals and Ethereum has a precision of 18 decimals. Is there support for 256 bit numbers?

**Comment by sarahdayan** [comment]
@joelnet Precision and exponent are different concepts. Bitcoins have an **exponent** of 8, while the issue here is allowing representation for "split sub-units", values that are virtually not billable (like $250.525) but may be needed for more precise calculations and or formatting purposes.

Bitcoins aren't ISO 4217, so they wouldn't be automatically supported as soon as #9 is taken care of. Yet, you'll be able to use `precision` to "bypass" that and create Dinero objects with bitcoins anyway by doing `Dinero({ amount: 1212345678, currency: 'XBT', precision: 8 })` 👍

**Comment by joelnet** [comment]
Thanks for the explanation. I have run into issues with precision before, I guess I never fully understood it until your explanation. I ended up solving it by increasing the exponent and then rounding for my final result.

Now that I understand the difference between precision and exponents, I need to go back in time and fix my code. Why can't I ever find my time machine when I need it most?

Cheers!

**Comment by scotttrinh** [comment]
@sarahdayan I definitely think backwards-compatibility is something to think about and would like to avoid breaking changes as much as possible! I think the basis of our disagreement is on what is the normal case:

1. fixed number of additional decimal places
2. fixed number of decimal places

The only evidence I have to support thinking Use 1 is the more common one is the idea of supporting fractional prices on bulk goods. You wouldn't want to encode the decimal places, since the extra precision might be truncated in the case of exponents greater than your usual. For example, you have an inventory application that allows subunit precision for bulk unit pricing, what would you put as `subUnitFractions` where it would be compatible with *any* exponent? Without being able to lookup the exponent you simply could not allow specifying additional precision.

For your example where the exponent changes to go beyond the specified precision (`0` to `3`), it would still break anyone who has serialized the `.value` instead of the `#toUnit()` value. Example:

```js
// JPY exponent: 0
const d = new Dinero({
  value: 1000,
  currency: 'JPY',
  subUnitFractions: 0
});
persistToStorage('price', d); // { value: 1000, currency: 'JPY', subUnitFractions: 0 }
d.toUnit(); // 1000

// Sometime in the future when the yen gets very valuable
// JPY exponent: 3
const dSpec = fromStorage('price'); // { value: 1000, currency: 'JPY', subUnitFractions: 0 }
const d = new Dinero(dSpec);
d.toUnit(); // 1
```

I still maintain that no one will be putting in non-"converted" values when creating a `Dinero` object. But, I would imagine that you would serialize the configuration object and reuse it and expect the same value, which is not the case in Solution 2.

FWIW, I think exponents changing is, unavoidably, a breaking change. But, in a sense, it's a breaking change for the entire financial system. We should *try* to consider how to best serve the users, but I think we can do that by providing helper functions to create the correct `.value` property from a string or a unit, and encourage the users to store the `#toUnit` or `#toFormat` value instead of anything that relies on the `.value` value.

**Comment by sarahdayan** [comment]
So, there's been quite some debate 🙂 I think at this point it's time to go back to the initial problem and simplify, so we can move on.

Currently, the notion of decimal places is hardcoded to `2`. It keeps people from accurately representing currencies that use less or more (e.g.: `IQD`, `JPY`), and it keeps people from representing monetary values with more precision than the currency allows by default (e.g. `$10.4545`).

We want to give more flexibility. In the end, people need to be able to indicate, for the amount they specify, what is the number of sub-units it represents (or if there are none). Whether you want to say `precision: 0` because you're using Yens, or because you want to manipulate mains units only, that's your business.

So, there should be **one concept**. Let's call it `precision` for now. This is the number of sub-units you want to represent. When you call `toUnit`, what gets returned is `amount / 10^precision`. It's simple, easy to understand, fits most cases.

Later, when a dataset is added with all the default exponents for each ISO 4217 currency, the default `2` will be replaced by the corresponding exponent to the currency you gave. So if you didn't specify a `precision`, this will be the new default value. If you did specify a `precision`, this overrides it. This is of course a breaking change, which is why it will only happen in v2.

But the bottom line is, you have the possibility to specify how many sub-units you're representing. Then it's all about what the default value is.

For more specific use cases, for example like the @scotttrinh gave where you might want to specify additional decimal places to the default ones depending on the currency, this can be handled by a converting method that adds or removes `precision` and changes the `amount` accordingly:

```
const d = Dinero({ amount: 100, currency: 'EUR' }) // precision isn't specified, so by default it's 2

d.addPrecision(3) // precision shifts to 5, amount shifts to 100000
d.removePrecision(4) // precision shifts to 1, amount shifts to 10

d.setPrecision(3) // precision is set to 3, amount remains untouched
```

This solution offers simplicity, reduces the risks for bugs, and allows to handle a lot of use cases.

**Comment by scotttrinh** [comment]
@sarahdayan 

I like the idea of having those precision shifting functions! To be clear, they wouldn't mutate, they'd return new `Dinero` objects, right? This definitely accomplishes the use-case I had in mind at the expense of creating the intermediate object, but that's a really tiny price to pay. Sounds like a good way forward!

I'll open a separate issue for the possible enhancement of adding deserializing (`#fromUnit` or `#fromFormat`) helper functions that might also include some of this functionality and is compatible with the idea of `precision` *always* indicating the decimal precision of the unit. I agree that is the most clear way to represent this.

**Comment by sarahdayan** [comment]
@scotttrinh Yes of course, always new objects. The library is and must remain immutable 🙂
